### ISSUE TITLE
Is it possible to limit the range of similarity computation? (Dynamic Time Warping) [K-Shape]

### ISSUE THREAD

**Issue by StefanR44** [tslearn-team/tslearn#370]
I would like to have DTW only move within a certain range, so that a point will not be matched to another point that is more than x steps away. 

I hope this visualization helps: 
[https://i.imgur.com/hlw9iOF.png](https://i.imgur.com/hlw9iOF.png)

Thanks in advance

Picture edited from Paparrizos and Gravano, "k-Shape: Efficient and Accurate Clustering of Time Series" (2016)

![DTW range limit](https://user-images.githubusercontent.com/90157240/143461905-e3baefb3-5103-4395-8dc3-9635e09a8210.png)
!

**Comment by GillesVandewiele** [comment]
That is most definitely possible. Please check our user guide: https://tslearn.readthedocs.io/en/stable/user_guide/dtw.html

The things you are looking for are called constraints. We support two types: `itakura` and `sakoe_chiba`

**Comment by StefanR44** [comment]
> That is most definitely possible. Please check our user guide: https://tslearn.readthedocs.io/en/stable/user_guide/dtw.html
> 
> The things you are looking for are called constraints. We support two types: `itakura` and `sakoe_chiba`

Ah, thanks a bunch @GillesVandewiele! I did not realize this was still k-shapes related. I will read into it!

**Comment by GillesVandewiele** [comment]
Most welcome. I am not sure about how to use it for K-shape myself at the moment. It might require some patching. In case you manage to make it work, please comment here on how you did it!

**Comment by StefanR44** [comment]
Will do! I will do some fiddling with K-shapes in some ways. The biggest issue with K-shape for me is, that it requires a dataset of equal-sized time series. For my use-case I am clustering time series of energy consumption and sometimes I do not have them for the whole year but only for a few months. 
Do you know by chance, if there is a way to do the clustering based on actual dates or months (e.g. summer series only matched with other summer series)?

**Comment by GillesVandewiele** [comment]
Not immediately from the top of my head... You could cut the longer timeseries into smaller windows as a preprocessing step or pad the shorter ones with noise perhaps? These are rather rudimentary and probably not the ideal solution though.

**Comment by StefanR44** [comment]
Yes, those are possible workarounds if everything else fails, I am actually on it right now 👍

**Comment by StefanR44** [comment]
Did not really succeed on that, as I realized that K-Shape does not use DTW but SBD. Now the questions is, how can that be adjusted? I was checking the functions for SBD usage, I guess this is where it is used / created. I tried to backtrack `y_shifted_sbd_vec`. It is a function from tslearn: `from tslearn.metrics import cdist_normalized_cc, y_shifted_sbd_vec`

Is there a way to find out more about it? When i wanted to go to the definition I just got this: 
`from .cycc import cdist_normalized_cc, y_shifted_sbd_vec`

What does it exactly do? 



    ```
def _shape_extraction(self, X, k):
        sz = X.shape[1]
        Xp = y_shifted_sbd_vec(self.cluster_centers_[k], X[self.labels_ == k],
                               norm_ref=-1,
                               norms_dataset=self.norms_[self.labels_ == k])
        S = numpy.dot(Xp[:, :, 0].T, Xp[:, :, 0])
        Q = numpy.eye(sz) - numpy.ones((sz, sz)) / sz
        M = numpy.dot(Q.T, numpy.dot(S, Q))
        _, vec = numpy.linalg.eigh(M)
        mu_k = vec[:, -1].reshape((sz, 1))

        # The way the optimization problem is (ill-)formulated, both mu_k and
        # -mu_k are candidates for barycenters
        # In the following, we check which one is best candidate
        dist_plus_mu = numpy.sum(numpy.linalg.norm(Xp - mu_k, axis=(1, 2)))
        dist_minus_mu = numpy.sum(numpy.linalg.norm(Xp + mu_k, axis=(1, 2)))
        if dist_minus_mu < dist_plus_mu:
            mu_k *= -1

        return mu_k
```

**Comment by jbbqqf** [comment]
Hi! Triaging older issues — I think this can be closed: range-limited DTW is already supported, just under a different name.

What you're describing is the Sakoe-Chiba band — capping `|i - j| ≤ r` in the DTW alignment matrix. `tslearn` exposes that via the `global_constraint=` and `sakoe_chiba_radius=` kwargs on `dtw` / `dtw_path` / `cdist_dtw`:

```python
from tslearn.metrics import dtw_path
path, dist = dtw_path(s1, s2, global_constraint="sakoe_chiba", sakoe_chiba_radius=3)
```

The mask itself is built by `tslearn.metrics.sakoe_chiba_mask` (see `tslearn/metrics/_masks.py:42`) and is the same constraint shown in the figure you posted from Paparrizos & Gravano. The Itakura parallelogram is also available with `global_constraint="itakura"` + `itakura_max_slope=`.

For `TimeSeriesKMeans(metric="dtw", metric_params={"global_constraint": "sakoe_chiba", "sakoe_chiba_radius": 3})` the same kwargs flow through `metric_params`. Would you mind closing if that covers your use case? Happy to reopen if you need the constraint somewhere it isn't currently plumbed.

(Disclosure: I drafted this comment with help from Claude Code while triaging stale issues; the references above were verified manually against `tslearn/metrics/_masks.py` and `dtw_variants.py`.)

**Comment by charavelg** [comment]
Close as answered
